### PR TITLE
[KCCM]:  have providerID trigger re-sync, but not be required for load balancer syncs

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

#117388 left some open ended questions around "are we sure all cloud providers even set a providerID at all?"...we are not. Should a cloud provider not set the providerID but still expect the service controller to provision LBs: then this won't work for them, because #117388 enforces that the providerID is defined on all nodes that we pass to the cloud provider when configuring load balancers. 

This PR revises that logic, and instead triggers a resync when any node gets the providerID assigned, but does not filter nodes based on it.    

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
[KCCM] drop filtering nodes for the providerID when syncing load balancers, but have changes to the field trigger a re-sync of load balancers. This should ensure that cloud providers which don't specify providerID, can still use the service controller implementation to provision load balancers.   
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/assign @thockin

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go [hunk 1]
```diff
@@ -978,16 +975,19 @@ func getNodePredicatesForService(service *v1.Service) []NodeConditionPredicate {
 	return allNodePredicates
 }
 
+// This predicate just validates if the providerID has been set and triggers a
+// node sync. It is _not_ used when determining which nodes to use when
+// configuring the load balancer's backend pool.
+func nodeHasProviderIDPredicate(node *v1.Node) bool {
```
Comment (thockin) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller.go:981]: We only use this in `shouldSyncUpdatedNode()` now, and we test whether the predicate changed.  IOW we handle "" to "value" and "value" to "", but not "foo" to "bar".  Should we just compare that ?

**Comment by aojea** [comment]
/assign @nckturner @andrewsykim 
seems we need to understand what are is the `providerID` contract with cloud-providers

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go (hunk 1 above)
Comment (alexanderConstantinescu) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:981]: Can the providerID even change? I was under the impression it can't

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Comment by alexanderKhaustov** [comment]
@alexanderConstantinescu 
Hi
There seems to be a bug here.
Please see the attached log. It demonstrates the following scenario
- a new node is added
- first updateLoadBalancerHosts happens, finishes successfully, updates lastSyncedNodes with a list including the new node
- new node gets providerID
- second updateLoadBalancerHosts happens, but since lastSyncedNodes already contains the new node, cloud provider's UpdateLoadBalancer is not triggered

So after all there's no UpdateLoadBalancer after setting node's providerID

[log1.txt](https://github.com/kubernetes/kubernetes/files/14209545/log1.txt)

**Comment by alexanderConstantinescu** [comment]
Hi @alexanderKhaustov 

Which version of Kube is this? You might be using a version which is missing: https://github.com/kubernetes/kubernetes/pull/120943

**Comment by alexanderKhaustov** [comment]
> Which version of Kube is this? You might be using a version which is missing: #120943

indeed, mine is before 1.28.4, thanks
Quite a dramatic story, split into 3 issues )

**Pull request merged**